compiler/optimizer/abstractinterpreter/InliningProposal.cpp

```diff
@@ -0,0 +1,304 @@
+
+   _cost = 0;
+   _benefit = 0;
+   }
+
+bool TR::InliningProposal::isEmpty()
+   {
+   if (!_nodes)
+      return true;
+
+   return _nodes->isEmpty();
+   }
+
+uint32_t TR::InliningProposal::getCost()
+   {
+   if (_cost == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _cost;
+   }
+
+double TR::InliningProposal::getBenefit()
+   {
+   if (_benefit == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _benefit;
+   }
+
+void TR::InliningProposal::computeCostAndBenefit()
+   {
+   _cost = 0;
+   _benefit = 0;
+
+   if (!_idt)
+      return;
+
+   TR_BitVectorIterator bvi(*_nodes);
+   int32_t idtNodeIndex;
+
+   while (bvi.hasMoreElements())
+      {
+      idtNodeIndex = bvi.getNextElement();
+      IDTNode *node = _idt->getNodeByGlobalIndex(idtNodeIndex - 1);
+      _cost += node->getCost();
+      _benefit += node->getBenefit();
+      }
+   }
+
+void TR::InliningProposal::ensureBitVectorInitialized()
+   {
+   TR_ASSERT_FATAL(!_frozen, "TR::InliningProposal::ensureBitVectorInitialized proposal is frozen, cannot be mutated");
+   if (!_nodes)
+      _nodes = new (_region) TR_BitVector(_region);
+   }
+
+bool TR::InliningProposal::isNodeInProposal(TR::IDTNode* node)
+   {
+   if (_nodes == NULL)
+      return false;
+   if (_nodes->isEmpty())
+      return false;
+
+   const int32_t idx = node->getGlobalIndex() + 1;
+
+   return _nodes->isSet(idx);
+   }
+
+void TR::InliningProposal::merge(TR::InliningProposal *a, TR::InliningProposal *b)
+   {
+   TR_ASSERT_FATAL(!_frozen, "TR::InliningProposal::merge proposal is frozen, cannot be mutated");
+   if (a->_nodes && b->_nodes)
+      {
+      TR_BitVector temp(*a->_nodes);
+      temp |= *b->_nodes;
+      ensureBitVectorInitialized();
+      *_nodes = temp;
+      }
+   else if (a->_nodes)
+      {
+      ensureBitVectorInitialized();
+      *_nodes = *a->_nodes;
+      }
+   else if (b->_nodes)
+      {
+      ensureBitVectorInitialized();
+      *_nodes = *b->_nodes;
+      }
+   else if (_nodes) // this need to be cleared out if both a->_nodes and b->_nodes are null
+      {
+      _nodes->empty();
+      }
+
+   _cost = 0;
+   _benefit = 0;
+   }
+
+bool TR::InliningProposal::intersects(TR::InliningProposal* other)
+   {
+   if (!_nodes || !other->_nodes)
+      return false;
+
+   return _nodes->intersects(*other->_nodes);
+   }
+
+
+TR::InliningProposalTable::InliningProposalTable(uint32_t rows, uint32_t cols, TR::Region& region) :
+      _rows(rows),
+      _cols(cols),
+      _region(region)
+   {
+   _table = new (region) InliningProposal**[rows];
+
+   for (uint32_t i = 0; i < rows; i ++)
+      {
+      _table[i] = new (region) InliningProposal*[cols]();
+      }
+   }
+
+TR::InliningProposal* TR::InliningProposalTable::getEmptyProposal()
+   {
+   static TR::InliningProposal* emptyProposal = new TR::InliningProposal(region(), NULL);
+
+   static std::once_flag flag;
+   std::call_once(flag, [] { emptyProposal->setFrozen(); });
```

Comment:
I believe we're supposed to avoid using the global `new` operator. Also, I'm not sure that we can count on having support for `call_once` or for lambdas. Finally, it feels a bit sketchy to me to pass `region()` into the constructor of an object that outlives the current compilation, even though I know that it won't be used for any allocations in this case

It should be sufficient to allocate one empty proposal per instance of `InliningProposalTable`. My original concern was more about allocating far more than necessary as the packing algorithm repeatedly calls `get()`

Reply:
I changed implementation for this part. Instrad of using static variable, I added another class member `_emptyProposal` for `InliningProposalTable`. It will initialize an empty proposal and frozen it when the `InliningProposalTable` is constructed. This implementation could allocate one empty proposal per instance of `InliningProposalTable` now.